docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1060 @@
+
+## Loading and storing vectors
+
+### Loading
+
+Both `Vector128` and `Vector256` provide at least five ways of loading them from memory:
+
+```cs
+public static class Vector128
+{
+    public static Vector128<T> Load<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadAligned<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadAlignedNonTemporal<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadUnsafe<T>(ref T source) where T : struct;
+    public static Vector128<T> LoadUnsafe<T>(ref T source, nuint elementOffset) where T : struct;
+}
+```
+
+The first three overloads require a pointer to the source. To be able to use a pointer in a safe way, the buffer needs to be pinned first (the GC is not tracking unmanaged pointers, we have to ensure that the memory does not get moved by GC in the meantime, as the pointers would silently become invalid). That is simple, the problem is doing the pointer arithmetic right:
+
+```cs
+unsafe int UnmanagedPointersSum(Span<int> buffer)
+{
+    fixed (int* pBuffer = buffer)
+    {
+        int* pEnd = pBuffer + buffer.Length;
+        int* pOneVectorFromEnd = pEnd - Vector128<int>.Count;
+        int* pCurrent = pBuffer;
+
+        Vector128<int> sum = Vector128<int>.Zero;
+
+        while (pCurrent <= pOneVectorFromEnd)
+        {
+            sum += Vector128.Load(pCurrent);
+
+            pCurrent += Vector128<int>.Count;
+        }
+
+        int result = Vector128.Sum(sum);
+
+        while (pCurrent < pEnd)
+        {
+            result += *pCurrent;
+
+            pCurrent++;
+        }
+
+        return result;
+    }
+}
+```
+
+The `LoadAligned` and `LoadAlignedNonTemporal` require the input to be aligned. Aligned reads and writes should be slightly faster but using them comes at a price of increased complexity.
+
+Currently .NET exposes only one API fo allocating unmanaged aligned memory: [NativeMemory.AlignedAlloc](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativememory.alignedalloc). In the future, we might provide [a dedicated API](https://github.com/dotnet/runtime/issues/27146) for allocating managed, aligned and hence pinned memory buffers.
+
+The alternative to creating aligned buffers (we don't always have the control over input) is to pin the buffer, find first aligned address, handle non-aligned elements, then start aligned loop and afterwards handle the remainder. Adding such complexity to our code is hardly ever worth it and needs to be proved with proper benchmarking on various hardware.
```

Comment:
> hardly ever worth it and needs to be proved

👍🏻

Reply:
We likely shouldn't have this wording at all.

Ensuring you process only aligned data may not be impactful to small amounts of data, but for medium to large data it can provide significant perf benefits as well as improving overall run to run stability and reducing noise.

All processors have some form of penalty for performing a read or write that splits a cache line or page boundary. The cost of this depends on the processor and the size/frequency of the split. Every architecture optimization manual goes into this extensively and explicitly recommends avoiding splits if possible.

As a summary, most modern processors have 64-byte cache lines and roughly 4kb pages. Split stores are almost universally more expensive than split loads. On older processors, the cost of a split page load was around 100 cycles and the cost of a split page store around 150 cycles. On more modern processors this is down to around 5 cycles for a split load (I don't recall the cost of a split page store). Likewise split cache lines loads use to cost around 20 cycles, that was reduced to 5 a few years back, and in the most recent down to a 1-2 cycle penalty.

These penalties can still add up and are much more prominent with wider vector sizes. For example, with `Vector128<T>` you'll (for misaligned data) hit a cache split every 4 operations and a page split every 256 operations, your `Vector256<T>` you'll hit a cache split every 2 operations and a page split every 128 operations, with `Vector512<T>` this again increases to a cache split every operation and a page split every 64 operations.

So for `Vector256<T>` you get 2x the cache split rate, and 4x for `Vector512<T>`. While some of this cost is mitigated by fewer instructions, you're still adding a non-trivial penalty to what is often one of the bigger (if not biggest) bottlenecks in many vectorized algorithms.